refactor!: convert from snafu to n0-error

Frando opened the pull request:
## Description

Changes the error handling across all crates in this repo from `snafu`/`n0-snafu` to **`n0-error`**. The latter is our new error library which features call-site location tracking and better ergonomics than snafu.

Repo: https://github.com/n0-computer/n0-error
Docs: https://docs.rs/n0-error/

*Note:*  The initial version of this PR was written by an AI (OpenAI codex) which I instructed in numerous prompts to convert iroh from n0-snafu/snafu to n0-error. It has since been cleaned up and reviewed line-by-line.

## Breaking Changes

*Changed*
* All error structs and enum variants now have a field `meta: n0_error::Meta`. Fields `backtrace` and `span_trace` are removed.
* All error structs and enums no longer implement `snafu::ErrorCompat`. They now implement `n0_error::StackError`. All our errors keep implementing `std::error::Error`. For downstream uses of these errors, not much changed.
* If you previously used `n0-snafu` to work with iroh errors, we recommend you switch to `n0-error`. See the [docs]( https://docs.rs/n0-error/) for how to work with n0-error.

## Notes & open questions

<!-- Any notes, remarks or open questions you have to make about the PR. -->

## Change checklist
<!-- Remove any that are not relevant. -->
- [x] Self-review.
- [x] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [x] Tests if relevant.
- [x] All breaking changes documented.
  - [ ] List all breaking changes in the above "Breaking Changes" section.
  - [ ] Open an issue or PR on any number0 repos that are affected by this breaking change. Give guidance on how the updates should be handled or do the actual updates themselves. The major ones are:
    - [ ] [`quic-rpc`](https://github.com/n0-computer/quic-rpc)
    - [ ] [`iroh-gossip`](https://github.com/n0-computer/iroh-gossip)
    - [ ] [`iroh-blobs`](https://github.com/n0-computer/iroh-blobs)
    - [ ] [`dumbpipe`](https://github.com/n0-computer/dumbpipe)
    - [ ] [`sendme`](https://github.com/n0-computer/sendme)

Frando committed:
refactor: convert iroh-base to n0-error

Frando committed:
fixup

Frando committed:
refactor: convert iroh-relay

Frando committed:
wip

Frando committed:
wip: converting iroh

Frando committed:
fmt

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
codex is converting iroh

Frando committed:
fixes

Frando committed:
wip

Frando committed:
codex did it

Frando committed:
fix & fmt

Frando committed:
chore: clippy fix & cargo fix & fmt

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
fixup

Frando committed:
fixup

dignifiedquire commented on iroh-base/src/key.rs:
```diff
@@ -384,14 +380,14 @@ fn decode_base32_hex(s: &str) -> Result<[u8; 32], KeyParsingError> {
         Ok(len) => {
             if len != PublicKey::LENGTH {
-                return Err(DecodeInvalidLengthSnafu.build());
+                return Err!(KeyParsingError::DecodeInvalidLength);
```
looking at this, I would much rather write

```rust
bail!(KeyParsingError::DecodeInvalidLength);
```

overloading `Err`  I find problematic personally

dignifiedquire replied:
I think that would be better to have than `Err!`

dignifiedquire commented on iroh-base/src/ticket.rs:
```diff
@@ -50,7 +49,7 @@ pub trait Ticket: Sized {
         let expected = Self::KIND;
         let Some(rest) = str.strip_prefix(expected) else {
-            return Err(KindSnafu { expected }.build());
+            return Err(ParseError::wrong_prefix(expected));
```
shouldn't this be called `ParseError::kind(expected)`?

dignifiedquire replied:
nvm, found the impl below

dignifiedquire replied:
could we have a macro so we can write

```rust
bail!(ParseError::Kind { expected });
```
without the  indirection method?

dignifiedquire commented on iroh-dns-server/src/config.rs:
```diff
@@ -173,7 +173,7 @@ impl Config {
         } else {
             let path = dirs_next::data_dir()
-                .context("operating environment provides no directory for application data")?;
+                .std_context("operating environment provides no directory for application data")?;
```
I think this works reasonably well, a bit confusing at first, but easy to use at the end of the day

Frando replied:
To me, this is the most annoying thing about this whole endevour, but unfortunately I don't think there's another solution. I made the `std_context` method the "longer" one (to type) to nudge towards using `context` wherever possible (ie. on our errors that implement StackError) because `std_context` will work on our errors too, but we'll lose the location info in the process. This is the much discussed issue that as long as we implement std::error::Error for our errors, we can't forbid calling the StdResultExt  methods on them, even though we *should*  always use the StackResultExt methods (which are only available on StackError results, not on std error results).

dignifiedquire commented on iroh-dns-server/src/http.rs:
```diff
@@ -173,11 +172,11 @@ impl HttpServer {
                 Ok(Err(err)) => {
                     warn!(?err, "task failed");
-                    final_res = Err(err).context("task");
+                    final_res = Err(err).std_context("task");
```
could we have something like 

```rust
final_res = whatever_std!(err, "task");
```

dignifiedquire commented on iroh-relay/src/client.rs:
```diff
@@ -220,10 +236,9 @@ impl ClientBuilder {
-                    url: dial_url.clone(),
-                }
-                .build()
+                e!(ConnectError::InvalidWebsocketUrl {
```
this macro fills in the location details?

dignifiedquire commented on iroh-relay/src/client/tls.rs:
```diff
@@ -144,14 +144,15 @@ impl MaybeTlsStreamBuilder {
         })
-        .await??;
+        .await
+        .map_err(|err| e!(DialError::Timeout, err))??;
```
what does this variant do?

dignifiedquire commented on iroh-relay/src/client/tls.rs:
```diff
@@ -240,22 +245,26 @@ impl MaybeTlsStreamBuilder {
-        let upgraded = hyper::upgrade::on(res).await.context(ProxyConnectSnafu)?;
+        let upgraded = hyper::upgrade::on(res)
+            .await
+            .map_err(|err| e!(DialError::ProxyConnect, err))?;
```
can't we do this with one of the `context` variants?

Frando replied:
`context` takes a string (or, impl Display) and adds this as context. We can't do without the `e!` macro if we want jump-to-usage on `ProxyConnect` to work and not type `meta: Default::default()` manually. So not sure what `context` should do here?

Frando commented on iroh-base/src/key.rs:
```diff
@@ -128,7 +127,7 @@ impl PublicKey {
         self.as_verifying_key()
             .verify_strict(message, &signature.0)
-            .map_err(|_| SignatureSnafu.build())
+            .map_err(|_| e!(SignatureError))
```
Right, this is a struct so it can just be `SignatureError::new()` in this case. However, for enums I removed the constructors from the proc macro generated code (I had them before) so that jump-to-definition on variants works.

Frando commented on iroh-relay/src/client/conn.rs:
```diff
@@ -146,9 +139,15 @@ impl Sink<ClientToRelayMsg> for Conn {
-        snafu::ensure!(size <= MAX_PACKET_SIZE, ExceedsMaxPacketSizeSnafu { size });
+        n0_error::ensure!(
+            size <= MAX_PACKET_SIZE,
+            n0_error::e!(SendError::ExceedsMaxPacketSize { size })
```
We can but then it wouldn't be usable with other error types anymore as it would add a `meta` field and if that doesn't exist it would give a compile error. We could add another macro `ensure_e!` or such, but not sure if that is really better than the current one. Will change to import `e` though, then it's much shorter already.

Frando committed:
update n0-error

Frando committed:
wip

Frando committed:
fix: adapt to n0-error API changes

Frando committed:
cleanup

Frando committed:
more cleanup

Frando committed:
cleanup

Frando committed:
updates and cleanups

Frando committed:
Merge remote-tracking branch 'origin/main' into Frando/n0-error

Frando committed:
self-review

Frando committed:
more self-review

Frando committed:
self review

Frando committed:
bump n0-error

Frando committed:
adapt for changes in n0-error

Frando committed:
refactor: bump n0-error, rename #display to #error

Frando committed:
wip: bail -> bail_any

Frando committed:
bump n0-error and adapt for changes

Frando committed:
bump n0-error

Frando committed:
bump n0-error, rename e() to anyerr()

Frando committed:
fixup

Frando committed:
bump n0-error

Frando committed:
Merge remote-tracking branch 'origin/main' into Frando/n0-error

Frando committed:
bump n0-error

Frando committed:
fixup

Frando committed:
bump n0-error

Frando committed:
bump to released n0-error

Frando committed:
deps: remove nested_enum_utils

dignifiedquire committed:
refactor: use ensure more

dignifiedquire committed:
refactor(iroh-dns-server): use concrete erros in utils

dignifiedquire commented on iroh-relay/src/client.rs:
```diff
@@ -272,10 +279,9 @@ impl ClientBuilder {
-                code: response.status(),
-            }
-            .fail()?;
+            return Err(e!(ConnectError::UnexpectedUpgradeStatus {
```
could use `ensure`

Frando committed:
fix: use imports more

Frando commented on iroh-dns-server/src/store/signed_packets.rs:
```diff
@@ -145,30 +145,32 @@ impl Actor {
+                                            tables.update_time.remove(&existing.timestamp().to_bytes(), key.as_bytes()).anyerr()?;
+                                            true
+                                        }
+                                    } _ => {
```
fixed

matheus23 commented on iroh/src/endpoint.rs:
```diff
@@ -2992,19 +2968,26 @@ mod tests {
+                .accept()
+                .anyerr()?
+                .await
+                .anyerr()?;
```
Can this be simplified to `server.accept().await.anyerr()?.await.anyerr()?;`?

While `Incoming::accept` returns a `Connecting`, `Incoming::into_future` turns it into an `IncomingFuture` which outputs `Connection`.

Frando committed:
refactor: simplify line in tests

Frando committed:
refactor(iroh-dns-server): move block from `select!` to fn (#3597)

## Description

`cargo fmt` can't handle code inside `tokio::select!` and the big match
statement had numerous formatting issues, so I moved it into a function.

## Breaking Changes

<!-- Optional, if there are any breaking changes document them,
including how to migrate older code. -->

## Notes & open questions

<!-- Any notes, remarks or open questions you have to make about the PR.
-->

## Change checklist
<!-- Remove any that are not relevant. -->
- [ ] Self-review.
- [ ] Documentation updates following the [style
guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text),
if relevant.
- [ ] Tests if relevant.
- [ ] All breaking changes documented.
- [ ] List all breaking changes in the above "Breaking Changes" section.
- [ ] Open an issue or PR on any number0 repos that are affected by this
breaking change. Give guidance on how the updates should be handled or
do the actual updates themselves. The major ones are:
    - [ ] [`quic-rpc`](https://github.com/n0-computer/quic-rpc)
    - [ ] [`iroh-gossip`](https://github.com/n0-computer/iroh-gossip)
    - [ ] [`iroh-blobs`](https://github.com/n0-computer/iroh-blobs)
    - [ ] [`dumbpipe`](https://github.com/n0-computer/dumbpipe)
    - [ ] [`sendme`](https://github.com/n0-computer/sendme)

Frando committed:
bump n0-watcher, iroh-metrics, and add git dep for portmapper and netwatch

Frando committed:
fix: less std errors

matheus23 commented on iroh-relay/src/server/http_server.rs:
```diff
@@ -474,41 +450,43 @@ impl RelayService {
 
         let upgrade_header = expect_header(&req, UPGRADE)?;
-        snafu::ensure!(
+        n0_error::ensure!(
```
Nit: A couple of qualified uses

Frando replied:
fixed

matheus23 commented on iroh/examples/transfer.rs:
```diff
@@ -212,7 +212,7 @@ impl EndpointArgs {
             #[cfg(not(feature = "test-utils"))]
             {
-                snafu::whatever!(
+                n0_error::bail_any!(
```
Nit: some more qualified uses

matheus23 reviewed (approved):
I only have small nits.

LGTM!

Frando committed:
update portmapper and netwatch

Frando commented on iroh-relay/src/client.rs:
```diff
@@ -271,12 +278,12 @@ impl ClientBuilder {
-        if response.status() != hyper::StatusCode::SWITCHING_PROTOCOLS {
-            UnexpectedUpgradeStatusSnafu {
-                code: response.status(),
+        n0_error::ensure!(
```
This function is `#[cfg(not(wasm_browser))]` and this is the only occurence of `ensure` in this module, so I left it like this instead of doing a conditional import of ensure at the module level.

Frando committed:
address some nits

Frando commented on iroh-relay/src/client.rs:
```diff
@@ -47,75 +46,85 @@ mod util;
 #[non_exhaustive]
 pub enum DialError {
-    #[snafu(display("Invliad target port"))]
+    #[error("Invliad target port")]
```
fixed

Frando commented on iroh-base/src/relay_url.rs:
```diff
@@ -54,7 +51,7 @@ impl FromStr for RelayUrl {
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let inner = Url::from_str(s).context(RelayUrlParseSnafu)?;
+        let inner = Url::from_str(s).map_err(|err| RelayUrlParseError::new(err))?;
```
fixed

Frando committed:
Merge remote-tracking branch 'origin/main' into Frando/n0-error

Frando committed:
Merge remote-tracking branch 'origin/main' into Frando/n0-error

Frando committed:
tests: mark integration_mainline as flaky
